:book: Migration doc (#895)
kubernetes-sigs/kubebuilder: docs/book/src/migration/v1vsv2.md

# Kubebuilder v1 vs v2

> This document should cover all breaking changes when migrating from v1 to v2.

Reviewer: ```suggestion
This document covers all breaking changes when migrating from v1 to v2.
```

(we can just be confident here :-P)

> The details of all changes can be found in

Reviewer: ```suggestion
The details of all changes (breaking or otherwise) can be found in the
```

[controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/releases),
[controller-tools](https://github.com/kubernetes-sigs/controller-tools/releases)
and [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/releases)
release notes.

## Common changes

> - All three projects now use `go modules` instead of `dep`.

Reviewer: ```suggestion
- All three projects now use `go modules` instead of `dep`.  Dep support in `controller-runtime` will be maintained until the release of Go 1.13 (when modules are enabled by default).
```

## controller-runtime

> - `client.List` now uses functional options `List(ctx, list, ...option)` instead

Reviewer: ```suggestion
- `Client.List` now uses functional options (`List(ctx, list, ...option)`) instead
```

of `List(ctx, ListOptions, list)`.

> - `client.DeleteAllOf` was added.

Reviewer: ```suggestion
- `Client.DeleteAllOf` was added to the `Client` interface.
```

- Metrics are on by default now.

- A number of packages under `pkg/runtime` have been moved, with their old

> locations deprecated. See the [godocs][pkg-runtime-godoc] for more information.

Reviewer: ```suggestion
locations deprecated.  The old locations will be removed before controller-runtime v1.0.0. See the [godocs][pkg-runtime-godoc] for more information.
```

#### Webhook-related

- Automatic certificate generation for webhooks has been removed, and webhooks
will no longer self-register. Use controller-tools to generate a webhook
configuration. If you need certificate generation, we recommend using
[cert-manager](https://github.com/jetstack/cert-manager). Kubebuilder v2 will
scaffold out cert manager configs for you to use -- see the
[Webhook Tutorial](/cronjob-tutorial/webhook-implementation.md) for more details.

> - The `builder` package now has separate builders for controllers and webhooks.

Reviewer: ```suggestion
- The `builder` package now has separate builders for controllers and webhooks, which facilitates choosing which to run.
```

## controller-tools

The generator framework has been rewritten in v2. It still works the same as

> before in many cases. But be aware that there are some breaking changes.

Reviewer: ```suggestion
before in many cases, but be aware that there are some breaking changes.
```

> Please check [marker document](/reference/markers.md) for more details.

Reviewer: ```suggestion
Please check [marker documentation](/reference/markers.md) for more details.
```

## kubebuilder

- Kubebuilder v2 introduces a simplified project layout. You can find the design
doc [here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/simplified-scaffolding.md).

> - In v1, the manager is deployed as `StatefulSet`, while as `Deployment` in v2.

Reviewer: ```suggestion
- In v1, the manager is deployed as a `StatefulSet`, while it's deployed as a `Deployment` in v2.
```

> - Add the `kubebuilder create webhook` command to scaffold

Reviewer: this construction makes it seem like you're telling the user to do something,  Use something like this instead:

```suggestion
- The `kubebuilder create webhook` command was added to scaffold
```

mutating/validating/conversion webhooks. It replaces the
`kubebuilder alpha webhook` command.

> - v2 uses `distroless/static` instead of ubuntu as base image.

Reviewer: ```suggestion
- v2 uses `distroless/static` instead of Ubuntu as base image.  This reduces image size and attack surface.
```

# Kubebuilder v1 vs v2

> There are a number of changes in

Reviewer: Lang: this feels a bit awkward or redundant.  I think this either needs to go away, or be replaced by a different intro.

[controller-runtime](https://github.com/kubernetes-sigs/controller-runtime),
[controller-tools](https://github.com/kubernetes-sigs/controller-tools)
and [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder).

> ## Common changes

Reviewer 2: We can probably rename this to `Project Level Changes` and merge this with the `Kubebuilder` changes listed at the bottom ?

Author: The original intend is to say the common changes in CR, CT and KB.

> - They have migrated from `dep` to `go modules` for managing dependencies.

Reviewer 2: We can say:
v2 uses `Go` modules instead `dep` ....

`They` reads a bit strange.

Author: I have rephrased `They` to make it less confusing.

> - They have migrated from `dep` to `go modules` for managing dependencies.

Reviewer: `s/They/All three projects/`

> ## Controller-runtime

Reviewer: nit: lowercase `c` since we always refer to `controller-runtime` as lowercase, like ee cummings :-P

> - Automatic certificate generation for webhooks has been removed, and webhooks

Reviewer 2: I think we can club webhook related changes together and move them below the `manager` and `client` changes because latter are more used more widely.

Author: Grouped webhook related stuff in a small section.

will no longer self-register. Use controller-tools to generate a webhook

> configuration, and use something like cert-manager to generate your certificates

Reviewer: Lang: split this sentence:

```markdown
Use controller-tools to generate a webhook configuration.  If you need certificate generation, we recommend
using [cert-manager](...).  KubeBuilder v2 will scaffold out cert manager configs for you to use -- see the 
[Webhook Tutorial](...) for more details.
```

> (if you don't already have a solution). Kubebuilder supports them out-of-the-box.

Reviewer: Lang: In `KubeBuilder supports them`, it's unclear what `them` is.  certmanager is a thing, so it should be `it`, if that's what you're talking about

> - If your API implements

Reviewer: We should probably only list changes that affect existing users.  Maybe add a note about that at the top.

Reviewer: Or maybe put new features elsewhere?  I don't want to have to update this doc every time something new comes out though.

[`webhook.Defaulter`](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/webhook#Defaulter)
and [`webhook.Validator`](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/webhook#Validator),
a webhook server will be automatically wired for you.
The tutorial can be found at [here](https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html).

> - Add CRD conversion webhook support. The tutorial can be found [here](https://github.com/kubernetes-sigs/kubebuilder/pull/851).

Reviewer 2: We should use the link to the tutorial in the book

> - `manager` now supports adding runables that is either managed by leader election

Reviewer: Lang: `runnables that are` -- plural --> are, singular --> is

Author: Dropped this item, since it's not considered as breaking change.

or always running. It depends on how runable implements
[LeaderElectionRunable](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/manager#LeaderElectionRunnable)
interface.

- `client` package:
  - `client.List` now uses functional options.

>   - Support delete collection.

Reviewer: Lang: this construction is awkward (also, perhaps confusingly, a singluar thing `supports`, plural things `support`)

>   - Add minimal server-side apply support. [#435](https://github.com/kubernetes-sigs/controller-runtime/pull/435)

Reviewer: ditto here

> - `builder` package now has 2 sets of builders for controller and webhook respectively.

Reviewer: Lang: respectively is generally used to note that two lists are related, but there's not two lists here.  Maybe `The builder package now has separate builders for controllers and webhooks.

(also plural controllers and webhooks, and you'll want to use articles)

- metrics are now on by default.

- A number of packages under pkg/runtime have been moved, with their old locations

> deprecated.

Reviewer: `... See the [godocs](...) for more information`

## Controller-tools

> The generator framework has been rewritten in v2. We tried our best to preserve the old behavior, but there some changes.

Reviewer: Did we? That was kind-of a non-goal in many cases for this particular change, and I don't want to send the wrong message.

This is actually important, since some markers changed how they behave.

Please check [../TODO.md] for more details.

## Kubebuilder

- Project structure has been simplified. You can find the design doc
[here](https://github.com/kubernetes-sigs/kubebuilder/blob/master/designs/simplified-scaffolding.md).

- v1 deploys manager as `StatefulSet`, while v2 deploys it as `Deployment`.

- add `kubebuilder create webhook` command to scaffold mutating/validating/conversion
webhooks.

- v1 uses ubuntu as base image for manager, v2 uses `distroless/static`.

> - v2 required kustomize v3.0.4+

Reviewer: Lang: requires (current-tense)

# Kubebuilder v1 vs v2

This document cover all breaking changes when migrating from v1 to v2.

The details of all changes (breaking or otherwise) can be found in
[controller-runtime](https://github.com/kubernetes-sigs/controller-runtime/releases),
[controller-tools](https://github.com/kubernetes-sigs/controller-tools/releases)
and [kubebuilder](https://github.com/kubernetes-sigs/kubebuilder/releases)
release notes.

## Common changes

> - All three projects now use `go modules` instead of `dep`. Dep support in

Reviewer 2: I think we can simplify this part when it comes to an end-user because they don't care if KB itself uses Go modules or not. We can simply say: 
"v2 project uses Go modules".

`controller-runtime` will be maintained until the release of Go 1.13 (when
modules are enabled by default).

> ## controller-runtime

Reviewer 2: Capitalize it ? `Controller runtime`

Author: Discussed offline and decided to keep as is.

- `Client.List` now uses functional options (`List(ctx, list, ...option)`) instead
of `List(ctx, ListOptions, list)`.
- `Client.DeleteAllOf` was added to the `Client` interface.

- Metrics are on by default now.

- A number of packages under `pkg/runtime` have been moved, with their old
locations deprecated. The old locations will be removed before
controller-runtime v1.0.0. See the [godocs][pkg-runtime-godoc] for more
information.

#### Webhook-related

- Automatic certificate generation for webhooks has been removed, and webhooks
will no longer self-register. Use controller-tools to generate a webhook
configuration. If you need certificate generation, we recommend using
[cert-manager](https://github.com/jetstack/cert-manager). Kubebuilder v2 will
scaffold out cert manager configs for you to use -- see the
[Webhook Tutorial](/cronjob-tutorial/webhook-implementation.md) for more details.

- The `builder` package now has separate builders for controllers and webhooks,
which facilitates choosing which to run.

> ## controller-tools

Reviewer 2: Capitalize it ? `Controller tools`

Author: Discussed offline and decided to keep as is.

The generator framework has been rewritten in v2. It still works the same as
before in many cases, but be aware that there are some breaking changes.

> Please check [marker documentation](/reference/markers.md) for more details.

Reviewer 2: `be aware` is a bit discouraging me to upgrade :)

Author: :rofl:

> ## kubebuilder

Reviewer 2: Capitalize it ? `Kubebuilder`

> - Kubebuilder v2 introduces a simplified project layout. You can find the design

Reviewer 2: s/simplified/simplified
